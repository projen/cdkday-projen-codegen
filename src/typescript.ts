import { FileBase, Project } from 'projen';
import { SourceCode } from './source-code';

export class TypeScriptSource extends SourceCode {
  constructor(project: Project, filePath: string) {
    super(project, filePath);

    // add the projen marker at the top followed by a newline
    this.comment(FileBase.PROJEN_MARKER);
    this.line();
  }

  public comment(text: string) {
    this.line(`// ${text}`);
  }

  public docstring(summary: string, options: DocstringOptions = {}) {
    this.line('/**');

    const emit = (s: string = '') => {
      for (const line of s.split('\n')) {
        this.line(` * ${line.trim()}`);
      }
    };

    emit(summary);

    if (options.remarks) {
      emit(); // newline before remarks
      emit(options.remarks);
    }

    const tags = options.tags ?? [];

    let defaultValue = options.defaultValue !== undefined ? JSON.stringify(options.defaultValue) : undefined;
    if (options.defaultDescription && defaultValue === undefined) {
      defaultValue = '-';
    }

    if (defaultValue) {
      tags.push({ default: `${defaultValue} ${options.defaultDescription ?? ''}` });
    }

    if (tags.length > 0) {
      emit(); // newline before tags
    }

    for (const tag of tags) {
      for (const [k, v] of Object.entries(tag)) {
        emit(`@${k} ${v.split('\n').join(' ')}`);
      }
    }

    this.line(' */');
  }
}

export interface DocstringOptions {
  readonly remarks?: string;

  /**
   * Adds a "@returns" tag.
   */
  readonly returns?: string;

  /**
   * Adds "@param" tags.
   */
  readonly params?: { [name: string]: string };

  /**
   * The default value (will be JSON stringified).
   *
   * If `defaultDescription` is specified and `defaultValue` is not specified, the default value will be "-".
   * @default - no default value
   */
  readonly defaultValue?: any;

  /**
   * The default description.
   * @default - no default description
   */
  readonly defaultDescription?: string;

  /**
   * Arbitrary tags to add to the docstring
   */
  readonly tags?: Array<{ [tag: string]: string }>;
}

export class TypeScriptInterface {

}