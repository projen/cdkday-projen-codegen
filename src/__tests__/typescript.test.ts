import { FileBase } from 'projen';
import { TypeScriptSource } from '../typescript';
import { TestProject } from './util';

test('typescript source includes the projen marker', () => {
  const project = new TestProject();
  new TypeScriptSource(project, 'hello.ts');
  expect(project.snapshot('hello.ts')).toStrictEqual(`// ${FileBase.PROJEN_MARKER}\n`);
});

test('line comment', () => {
  const project = new TestProject();
  const ts = new TypeScriptSource(project, 'hello.ts');

  ts.comment('another comment');
  ts.open();
  ts.comment('indented comment');
  ts.close();

  expect(project.snapshot('hello.ts')).toMatchSnapshot();
});

describe('docstring', () => {
  test('summary only', () => {
    const project = new TestProject();
    const ts = new TypeScriptSource(project, 'hello.ts');

    ts.docstring('summary');

    expect(project.snapshot('hello.ts')).toMatchSnapshot();
  });

  test('multiline summary', () => {
    const project = new TestProject();
    const ts = new TypeScriptSource(project, 'hello.ts');

    ts.docstring('summary\nline2\nline3');

    expect(project.snapshot('hello.ts')).toMatchSnapshot();
  });

  test('summary + remarks', () => {
    const project = new TestProject();
    const ts = new TypeScriptSource(project, 'hello.ts');

    ts.docstring('summary', {
      remarks: 'Hello, world this is awesome\nanother remakrs section',
    });

    expect(project.snapshot('hello.ts')).toMatchSnapshot();
  });

  test('tags', () => {
    const project = new TestProject();
    const ts = new TypeScriptSource(project, 'hello.ts');

    ts.docstring('summary\nline2\nline3', {
      tags: [
        { param: 'foo\tbar' },
        {
          returns: 'Hello world',
          default: 'boom boom',
        },
        { param: 'zoo\tzar\nnewlinx' },
      ],
    });

    expect(project.snapshot('hello.ts')).toMatchSnapshot();
  });

  test('default value', () => {
    const project = new TestProject();
    const ts = new TypeScriptSource(project, 'hello.ts');

    ts.docstring('string', {
      defaultValue: 'this is the default value',
    });

    ts.docstring('boolean', {
      defaultValue: false,
      defaultDescription: 'no no no',
    });

    ts.docstring('number', {
      defaultValue: 1234,
    });

    ts.docstring('array', {
      defaultValue: [1, 2, 3, 'hello'],
      defaultDescription: 'this is an array',
    });

    ts.docstring('just description', {
      defaultDescription: 'no default value but there is a desc',
    });

    expect(project.snapshot('hello.ts')).toMatchSnapshot();
  });
});