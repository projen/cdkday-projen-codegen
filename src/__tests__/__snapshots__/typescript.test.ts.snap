// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docstring default value 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

/**
 * string
 * 
 * @default \\"this is the default value\\"
 */
/**
 * boolean
 * 
 * @default false no no no
 */
/**
 * number
 * 
 * @default 1234
 */
/**
 * array
 * 
 * @default [1,2,3,\\"hello\\"] this is an array
 */
/**
 * just description
 * 
 * @default - no default value but there is a desc
 */"
`;

exports[`docstring multiline summary 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

/**
 * summary
 * line2
 * line3
 */"
`;

exports[`docstring summary + remarks 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

/**
 * summary
 * 
 * Hello, world this is awesome
 * another remakrs section
 */"
`;

exports[`docstring summary only 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

/**
 * summary
 */"
`;

exports[`docstring tags 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

/**
 * summary
 * line2
 * line3
 * 
 * @param foo	bar
 * @returns Hello world
 * @default boom boom
 * @param zoo	zar newlinx
 */"
`;

exports[`line comment 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

// another comment
  // indented comment"
`;
